### PR TITLE
Add Protobuf v23.1

### PR DESCRIPTION
**Pull request by mering**
Add `protobuf` `v23.1`.

This also adds `upb` `0.0.0-20230516-61a97ef` https://github.com/protocolbuffers/upb/commit/61a97efa24a5ce01fb8cc73c9d1e6e7060f8ea98 which cyclic depends with protobuf.

### CODE REVIEW

**Comment by fmeum** [comment]
Thank you very much for working on this, it will unblock a migration I had planned for a while now.

Looks like upb is missing a patch to bring in the module file.

**Comment by mering** [comment]
@fmeum There seems to be something wrong with the test:

```
cd modules/upb/0.0.0-20230516-61a97ef
mkdir test && cd test && git init
git apply ../patches/module_dot_bazel.patch ../patches/utf8_range.patch
diff MODULE.bazel ../MODULE.bazel
```

doesn't report any changes between the `MODULE.bazel` checked in and the one created with the patches.

File: modules/upb/0.0.0-20230516-61a97ef/source.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "integrity": "sha256-fRnyrJweUIqGonKRPZqmfIFHgn+UkDWCiRC7BdnyzwM=",
+    "strip_prefix": "upb-61a97efa24a5ce01fb8cc73c9d1e6e7060f8ea98",
+    "url": "https://github.com/protocolbuffers/upb/archive/61a97efa24a5ce01fb8cc73c9d1e6e7060f8ea98.tar.gz",
+    "patch_strip": 0,
```
Comment (Wyverald) [inline modules/upb/0.0.0-20230516-61a97ef/source.json:5]: you need patch_strip 1 here

Comment (mering) [reply modules/upb/0.0.0-20230516-61a97ef/source.json]: Thanks!

**Comment by mering** [comment]
Marking as draft as I don't understand the failures in the tests. Any help appreciated.

File: modules/protobuf/23.1/presubmit.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+matrix:
+  platform: ["centos7", "debian10", "macos", "ubuntu2004", "windows"]
```
Comment (fmeum) [inline modules/protobuf/23.1/presubmit.yml:2]:
```suggestion
  platform: ["debian10", "macos", "ubuntu2004", "windows"]
```
Just drop centos7, it's way too old.

Comment (mering) [reply modules/protobuf/23.1/presubmit.yml:2]: Thanks, done.

**Comment by fmeum** [comment]
Looks like this variable (https://github.com/protocolbuffers/upb/blob/e7430e66d6e51def2a88f0b66fdab62b0d9492c1/upbc/bootstrap_compiler.bzl#L23) somehow ends up pointing to `external/com_google_protobuf` rather than `external/protobuf~...`. Not sure why though.

**Comment by mering** [comment]
> Looks like this variable (https://github.com/protocolbuffers/upb/blob/e7430e66d6e51def2a88f0b66fdab62b0d9492c1/upbc/bootstrap_compiler.bzl#L23) somehow ends up pointing to `external/com_google_protobuf` rather than `external/protobuf~...`. Not sure why though.

@fmeum Thanks, this was the hint I needed! It was hardcoded in the used version so I backported this line from the latest one.

**Review by fmeum — APPROVED** [review]

**Pull request merged**